specs/sharding/beacon-chain.md

```diff
@@ -103,786 +101,481 @@ The following values are (non-configurable) constants used throughout the specif
+### Extended Containers
 
-The commitments are not further collapsed to a single hash, 
-to avoid an extra network roundtrip between proposer and builder, to include the header on-chain more quickly.
+#### `BeaconState`
 
 ```python
-class ShardBlobBodySummary(Container):
-    # The actual data commitment
-    commitment: DataCommitment
-    # Proof that the degree < commitment.samples_count * POINTS_PER_SAMPLE
-    degree_proof: BLSCommitment
-    # Hash-tree-root as summary of the data field
-    data_root: Root
-    # fee payment fields (EIP 1559 like)
-    # TODO: express in MWei instead?
-    max_priority_fee_per_sample: Gwei
-    max_fee_per_sample: Gwei
+class BeaconState(bellatrix.BeaconState):
+    blocks_since_intermediate_block: List[BeaconBlock, MAX_BEACON_BLOCKS_BETWEEN_INTERMEDIATE_BLOCKS]
 ```
 
-### `ShardBlob`
-
-`ShardBlobBody` wrapped with the header data that is unique to the shard blob proposal.
+#### `BeaconBlockBody`
 
 ```python
-class ShardBlob(Container):
-    slot: Slot
-    shard: Shard
-    # Builder of the data, pays data-fee to proposer
-    builder_index: BuilderIndex
-    # Proposer of the shard-blob
-    proposer_index: ValidatorIndex
-    # Blob contents
-    body: ShardBlobBody
+class BeaconBlockBody(bellatrix.BeaconBlockBody):
+    execution_payload: Union[None, ExecutionPayload]
+    sharded_commitments_container: Union[None, ShardedCommitmentsContainer]
+    intermediate_block_bid: Union[None, IntermediateBlockBid]
 ```
 
-### `ShardBlobHeader`
-
-Header version of `ShardBlob`.
+## Helper functions
 
-```python
-class ShardBlobHeader(Container):
-    slot: Slot
-    shard: Shard
-    # Builder of the data, pays data-fee to proposer
-    builder_index: BuilderIndex
-    # Proposer of the shard-blob
-    proposer_index: ValidatorIndex
-    # Blob contents, without the full data
-    body_summary: ShardBlobBodySummary
-```
+*Note*: The definitions below are for specification purposes and are not necessarily optimal implementations.
 
-### `SignedShardBlob`
+### Block processing
 
-Full blob data, signed by the shard builder (ensuring fee payment) and shard proposer (ensuring a single proposal).
+#### `is_intermediate_block_slot`
 
 ```python
-class SignedShardBlob(Container):
-    message: ShardBlob
-    signature: BLSSignature
+def is_intermediate_block_slot(slot: Slot):
+    return slot % 2 == 1
 ```
 
-### `SignedShardBlobHeader`
+### KZG
 
-Header of the blob, the signature is equally applicable to `SignedShardBlob`.
-Shard proposers can accept `SignedShardBlobHeader` as a data-transaction by co-signing the header.
+#### `hash_to_field`
 
 ```python
-class SignedShardBlobHeader(Container):
-    message: ShardBlobHeader
-    # Signature by builder.
-    # Once accepted by proposer, the signatures is the aggregate of both.
-    signature: BLSSignature
+def hash_to_field(x: Container):
+    return int.from_bytes(hash_tree_root(x), "little") % MODULUS
```

Comment:
The resulting values would not have uniform distribution: lower values are more likely to occur. 
Could it potentially have any security impact? Or collision resistance is sufficient property?

Reply:
We do need a random oracle, as this is mainly used for Fiat-Shamir constructions. However, I believe the impact of a 1 bit bias is not important. This can be confirmed by the cryptography team.